### PR TITLE
[IMP] event_sale: add event on sale order

### PR DESCRIPTION
**Pull request by angv-odoo**
[IMP] event_sale: add event on sale order

This commit allows us to add event on sale order in an order line, as well as adding events via template.

Ticket-2988003

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        res = self._convert_to_write(res)
+        return res
+
+    def action_configure(self):
+        self.ensure_one()
+        for configurable_event_id in self.configurable_event_ids:
+            configurable_event_id.get_event_data()
+        return {'type': 'ir.actions.act_window_close'}
+
+class EventConfiguratorLine(models.TransientModel):
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:41]: can we move this model in a different file ?

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    """Event Configuration"""
+    _name = "event.editor.line"
+    _description = 'Edit Event Lines on Sales Confirmation'
+    _order = "id desc"
+
+    editor_id = fields.Many2one('event.event.configurator.batch')
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:47]: miss a field label

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @api.model
+    def default_get(self, fields):
+        res = super(BatchEventConfigurator, self).default_get(fields)
+        if not res.get('sale_order_id'):
+            sale_order_id = res.get('sale_order_id', self._context.get('active_id'))
+            res['sale_order_id'] = sale_order_id
+        sale_order = self.env['sale.order'].browse(res.get('sale_order_id'))
+        configurable_event_ids = []
+        for so_line in [l for l in sale_order.order_line if l.product_type == 'event']:
+            for _ in range(int(so_line.product_uom_qty)):
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:23]: use double underscore for unused variable (otherwise you will shadow the `_` used for translation)
```suggestion
            for __ in range(int(so_line.product_uom_qty)):
```

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+class BatchEventConfigurator(models.TransientModel):
+    _name = 'event.event.configurator.batch'
+    _description = 'Event batch Configurator'
+
+    configurable_event_ids = fields.One2many('event.editor.line', 'editor_id')
+    sale_order_id = fields.Many2one('sale.order', 'Sales Order', required=True, ondelete='cascade')
+
+    @api.model
+    def default_get(self, fields):
+        res = super(BatchEventConfigurator, self).default_get(fields)
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:16]:
```suggestion
        res = super().default_get(fields)
```

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models, fields
+
+
+class BatchEventConfigurator(models.TransientModel):
+    _name = 'event.event.configurator.batch'
+    _description = 'Event batch Configurator'
+
+    configurable_event_ids = fields.One2many('event.editor.line', 'editor_id')
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:11]: miss a field label

File: addons/event_booth_sale/wizard/event_booth_batch_configurator.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                value = {
+                    'event_id': self.event_id.id,
+                    'order_id': self.order_id.id,
+                    'event_booth_ids': self.event_booth_ids.ids,
+                    'event_booth_category_id': self.event_booth_category_id.id,
+                    'display_type': line.display_type,
+                    'product_id' : line.product_id.id,
+                    'product_uom' : line.product_uom.id,
+                }
+                line.write(value)
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator.py:82]: no need to create a `value` variable, you can directly write the dict value

File: addons/event_booth_sale/security/ir.model.access.csv [hunk 7]
```diff
@@ -3,3 +3,5 @@ access_event_booth_registration_salesman,access.event.booth.registration.salesma
 access_event_booth_registration_event_desk,access.event.booth.registration.event.desk,model_event_booth_registration,event.group_event_registration_desk,1,0,0,0
 access_event_booth_registration_event_user,access.event.booth.registration.event.user,model_event_booth_registration,event.group_event_user,1,1,1,1
 access_event_booth_configurator_salesman,access.event.booth.configurator.salesman,model_event_booth_configurator,sales_team.group_sale_salesman,1,1,1,0
+access_event_event_booth_configurator_batch,access.event.event.booth.configurator.batch,model_event_event_booth_configurator_batch,sales_team.group_sale_salesman,1,1,1,0
+access_event_booth_editor_line,access.event.booth.editor.line,model_event_booth_editor_line,sales_team.group_sale_salesman,1,1,1,1
```
Comment (std-odoo) [inline addons/event_booth_sale/security/ir.model.access.csv:7]: missing a \n at the end of the file

File: addons/event_booth_sale/wizard/event_booth_batch_configurator.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models, fields
+
+
+class BoothBatchEventConfigurator(models.TransientModel):
+    _name = 'event.event.booth.configurator.batch'
+    _description = 'Event booth batch Configurator'
+
+    configurable_booth_event_ids = fields.One2many('event.booth.editor.line', 'editor_id')
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator.py:11]: same here

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

First batch of comments, thanks for your work :)

File: addons/event_sale/wizard/event_batch_configurator.py (hunk 1 above)
Comment (angv-odoo) [reply addons/event_sale/wizard/event_batch_configurator.py:41]: We can, but then we would want to move the model from event_edit_registration.py also, otherwise I think it won't really be consistent :/

Comment (std-odoo) [reply addons/event_sale/wizard/event_batch_configurator.py:41]: yes we can move it as well :)

File: addons/event_sale/wizard/event_batch_configurator.py (hunk 2 above)
Comment (angv-odoo) [reply addons/event_sale/wizard/event_batch_configurator.py]: Not sure if there is a reason for this field to have a field label, considering that we are not going to want to show this value ever in front-end, and as in other wizards of the module the editor_id fields do not have labels either. 

If you still think I should add field label here, maybe we should also update other configurator id fields in the event_Sale and event_booth_sale module wizards?

Comment (std-odoo) [reply addons/event_sale/wizard/event_batch_configurator.py]: I would not change existing wizard, but generally it's a good practise to have field label even for technical field

File: addons/event_booth_sale/wizard/event_booth_batch_configurator.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            for _ in range(int(so_line.product_uom_qty)):
+                configurable_booth_event_ids.append([0, 0, {
+                    'event_id': so_line.event_id.id,
+                    'event_booth_ids': so_line.event_booth_ids.ids,
+                    'event_booth_category_id': so_line.event_booth_category_id.id,
+                    'product_id': so_line.product_id,
+                    'order_id': sale_order,
+                    'sale_order_line_id':so_line,
+                }])
+        res['configurable_booth_event_ids'] = configurable_booth_event_ids
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator.py:32]: list comprehension is better :) 

Also, try to not use `_` as a variable because it's used fr translation as well, always prefer `__` for unused variable
```suggestion
        res['configurable_booth_event_ids'] = [
            [0, 0, {
                'event_id': so_line.event_id.id,
                'event_booth_ids': so_line.event_booth_ids.ids,
                'event_booth_category_id': so_line.event_booth_category_id.id,
                'product_id': so_line.product_id,
                'order_id': sale_order,
                'sale_order_line_id':so_line,
            }]
            for so_line in sale_order.order_line
            for __ in range(int(so_line.product_uom_qty))
            if so_line.product_type == 'event_booth'
        ]
```

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    def default_get(self, fields):
+        res = super().default_get(fields)
+        if not res.get('sale_order_id'):
+            sale_order_id = res.get('sale_order_id', self._context.get('active_id'))
+            res['sale_order_id'] = sale_order_id
+        sale_order = self.env['sale.order'].browse(res.get('sale_order_id'))
+        configurable_event_ids = []
+        for so_line in [l for l in sale_order.order_line if l.product_type == 'event']:
+            for __ in range(int(so_line.product_uom_qty)):
+                configurable_event_ids.append([0, 0, {
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:24]: same here about list comprehension

File: addons/event_sale/wizard/event_batch_configurator_line.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        self.ensure_one()
+        for line in self.sale_order_line_id:
+            if line.product_type == 'event' and not line.event_id.id:
+                line.write({
+                    'event_id': self.event_id.id,
+                    'event_ticket_id': self.event_ticket_id.id,
+                    'order_id': self.order_id.id,
+                    'display_type': line.display_type,
+                    'product_id' : line.product_id.id,
+                    'product_uom' : line.product_uom.id,
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator_line.py:30]: why is this needed ? you rewrite the same value isn't it ? (display_type / product_id / product_uom)

is not necessary, I think the write can be done in batch

File: addons/event_sale/wizard/event_edit_registration_line.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_edit_registration_line.py:1]: since this is only code move, can we do it in a separate commit ?

File: addons/event_sale/wizard/event_batch_configurator_line.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    _order = "id desc"
+
+    editor_id = fields.Many2one('event.event.configurator.batch')
+    product_id = fields.Many2one('product.product', string="Product", readonly=True)
+    event_ticket_id = fields.Many2one('event.event.ticket', string='Event Ticket')
+    event_id = fields.Many2one('event.event', string="Event")
+    order_id = fields.Many2many('sale.order', string="Sale order")
+    sale_order_line_id = fields.Many2many('sale.order.line', string="Sale Order Line")
+
+    def get_event_data(self):
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator_line.py:20]: no need it to be public (public == you can call it remotely in RPC)

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    def get_event_data(self):
+        self.ensure_one()
+        for line in self.sale_order_line_id:
+            if line.product_type == 'event_booth' and not line.event_id.id:
+                line.write({
+                    'event_id': self.event_id.id,
+                    'order_id': self.order_id.id,
+                    'event_booth_ids': self.event_booth_ids.ids,
+                    'event_booth_category_id': self.event_booth_category_id.id,
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py:42]: this is not extensible, you overwrite the entire method to write 2 other field, "event_booth_ids" and "event_booth_category_id", but what if one day we have a different module depending on event_sale but not on event_booth_sale ?

one module will overwrite the other and some field will be missing

instead what you can do is a prepare method that return the value to write, and you call the super and concatenate the "super values" with the current one

something like
```python
def _prepare_event_values(self):
    return {
        ** super()._prepare_event_values()
        "event_booth_ids": ...,
        "event_booth_category_id": ...,
    }
```

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    event_booth_category_available_ids = fields.Many2many(related='event_id.event_booth_category_available_ids', readonly=True)
+    order_id = fields.Many2many('sale.order', string="Sale order")
+    sale_order_line_id = fields.Many2many('sale.order.line', string="Sale Order Line")
+
+    @api.depends('event_id')
+    def _compute_event_booth_category_id(self):
+        self.event_booth_category_id = False
+
+    @api.depends('event_id', 'event_booth_category_id')
+    def _compute_event_booth_ids(self):
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py:31]: can we add a docstring to explain it act like an onchange ? (the objective is just to reset the value if we change the event / category)

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

Thanks for the update, I have a few more comments :) 

Cheers

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py (hunk 14 above)
Comment (angv-odoo) [reply addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py:42]: Hey, thanks for review! I have a small question regarding this part:

Since we are creating a new model for the wizard and not inheriting one, it will not override anything even if we have similar function names no? Thus, if someone will want their module to depend on event_sale and not on event_sale_booth, they will be changing _get_event_data function of "event.editor.line" model, not the _get_event_data function of "event.booth.editor.line" model. Am I missing something in the logic here?

Comment (std-odoo) [reply addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py:42]: ah yes sorry, you are right :) mistake from me, sorry

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    def _get_event_data(self):
+        self.ensure_one()
+        for line in self.sale_order_line_id:
+            if line.product_type == 'event_booth' and not line.event_id.id:
+                line.write({
+                    'event_id': self.event_id.id,
+                    'order_id': self.order_id.id,
+                    'event_booth_ids': self.event_booth_ids.ids,
+                    'event_booth_category_id': self.event_booth_category_id.id,
+                })
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_line.py:45]: the write can be batched


```suggestion
        lines = self.sale_order_line_id.filtered(
            lambda line: line.product_type == 'event_booth' and not line.event_id.id)
        lines.write({
                'event_id': self.event_id.id,
                'order_id': self.order_id.id,
                'event_booth_ids': self.event_booth_ids.ids,
                'event_booth_category_id': self.event_booth_category_id.id,
            })
```

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+        sale_order = self.env['sale.order'].browse(res.get('sale_order_id'))
+        res['configurable_event_ids'] = [
+            [0, 0, {
+                    'event_id': so_line.event_id.id,
+                    'event_ticket_id': so_line.event_ticket_id.id,
+                    'product_id': so_line.product_id,
+                    'order_id': sale_order,
+                    'sale_order_line_id':so_line,
+                }]
+        for so_line in sale_order.order_line
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:29]: same here about indentation

File: addons/event_sale/wizard/event_batch_configurator_line.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    sale_order_line_id = fields.Many2many('sale.order.line', string="Sale Order Line")
+
+    @api.depends('event_id')
+    def _compute_event_ticket_ids(self):
+        # Once event is changed, ticket category should reset
+        self.event_ticket_id = False
+
+    def _get_event_data(self):
+        self.ensure_one()
+        for line in self.sale_order_line_id:
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator_line.py:28]: same here about batch

File: addons/event_sale/wizard/event_edit_registration_line.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    sale_order_line_id = fields.Many2one('sale.order.line', string='Sales Order Line')
+    event_id = fields.Many2one('event.event', string='Event', required=True)
+    registration_id = fields.Many2one('event.registration', 'Original Registration')
+    event_ticket_id = fields.Many2one('event.event.ticket', string='Event Ticket')
+    email = fields.Char(string='Email')
+    phone = fields.Char(string='Phone')
+    mobile = fields.Char(string='Mobile')
+    name = fields.Char(string='Name')
+
+    def get_registration_data(self):
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_edit_registration_line.py:21]: can we make it private while we change the file ?

File: addons/event_sale/wizard/event_edit_registration_line.py (hunk 12 above)
Comment (std-odoo) [inline addons/event_sale/wizard/event_edit_registration_line.py:1]: missing the license line
```
# Part of Odoo. See LICENSE file for full copyright and licensing details.
```

File: addons/event_booth_sale/wizard/event_booth_batch_configurator.py [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+                'product_id': so_line.product_id,
+                'order_id': sale_order,
+                'sale_order_line_id':so_line,
+            }]
+        for so_line in sale_order.order_line
+        for __ in range(int(so_line.product_uom_qty))
+        if so_line.product_type == 'event_booth'
+        ]
+        res = self._convert_to_write(res)
+        return res
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator.py:35]: no need to create a `res` variable, we can directly return

File: addons/event_sale/wizard/event_batch_configurator.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        for __ in range(int(so_line.product_uom_qty))
+        if so_line.product_type == "event"
+        ]
+        res = self._convert_to_write(res)
+        return res
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator.py:34]: same here

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+                                ('state', '=', 'available')
+                           ]"/> 
+                    <field name="order_id" invisible="1"/>
+                    <field name="sale_order_line_id" invisible="1"/>
+
+                            </tree>
+                        </field>
+                    </sheet>
+                    <footer>
+                        <button string="Save Configuration" name="action_configure" type="object" class="btn-primary" data-hotkey="q"/>
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml:34]: can be multilined

File: addons/event_sale/wizard/event_batch_configurator_views.xml [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+                    />
+                    <field
+                        name="event_ticket_id"
+                        domain="[('event_id', '=', event_id), ('product_id', '=', product_id)]"
+                        attrs="{
+                            'required': [('event_id', '!=', False)],
+                        }"
+                        context="{'name_with_seats_availability': True}"
+                        options="{'no_open': True, 'no_create': True}"
+                    />
```
Comment (std-odoo) [inline addons/event_sale/wizard/event_batch_configurator_views.xml:30]: wrong indentation

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+                           domain="[('id', 'in', event_booth_category_available_ids)]"/> 
+                    <field name="event_booth_ids" widget="many2many_tags"
+                           domain="[
+                                ('event_id', '=', event_id),
+                                ('booth_category_id', '=', event_booth_category_id),
+                                ('product_id', '=', product_id),
+                                ('state', '=', 'available')
+                           ]"/> 
+                    <field name="order_id" invisible="1"/>
+                    <field name="sale_order_line_id" invisible="1"/>
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml:28]: same here

File: addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (std-odoo) [inline addons/event_booth_sale/wizard/event_booth_batch_configurator_views.xml:29]: empty line

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

Thanks for the update, I have a few more comments and we should be good :)
